### PR TITLE
[stable/redis-ha] Change to statefulset (from #4038) 

### LINKED ISSUE #3403: [stable/redis-ha] PVC possibility was removed, why?
By looking at git history I can just see "Removing PVC per PR". It is unclear however what the motivation for this was or what users can do about it. I would like to mount a persistent volume for redis data. So I am tempted to bring back the old commit in my fork, but of course I have to ask first what the motivation for this change on stable was in first place.  

Yes, redis is usually used only as a cache and you don't need persistence. And also if you do, you should store an external .rdb backup or something. Still there are cases where persistent volume might makes sense. (I guess it is unnecessary to dive into the specifics on my use-case here)

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Feature request (I suppose)

**Version of Helm and Kubernetes**:
Helm: Client: &version.Version{SemVer:"v2.7.2", GitCommit:"8478fb4fc723885b155c924d1c8c410b7a9444e6", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.7.2", GitCommit:"8478fb4fc723885b155c924d1c8c410b7a9444e6", GitTreeState:"clean"}
Kubernetes: Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.5", GitCommit:"cce11c6a185279d037023e02ac5249e14daa22bf", GitTreeState:"clean", BuildDate:"2017-12-07T16:05:18Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

**Which chart**:
stable/redis-ha

**What happened**:
PVC feature was present, but got removed in last version

**What you expected to happen**:
Alternative mentioned in documentation or PVC configuration still existing

**How to reproduce it** (as minimally and precisely as possible):
helm install stable/redis-ha or just looking into values.yaml

### PR DESCRIPTION
**Pull request by RaymondArias**
**What this PR does / why we need it**:
The pull request converts the redis-ha chart back to a statefulset.

It is essentially a copy of #4038 with the changes requested in the pull requests. Also fixed a service account issue. 
I created this since @bzumhagan has gone MIA for two months and this change is very much needed for this chart.
@smileisak 
@ianmaddox 
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #3403 

**Special notes for your reviewer**:
I feel this change is needed since the current redis-ha chart has no persistent storage. Which is important if the redis-server pods go down.

### CODE REVIEW

**Comment by dung95** [comment]
please check to soon! I need to this

**Comment by RaymondArias** [comment]
I really want this merged as well....this chart desperately needs this pr for the ability to use persistent storage later on

**Comment by shantanuthatte** [comment]
👍 Waiting on this.

**Comment by RaymondArias** [comment]
@unguiculus and @mgoodness could you please review this pull request? My company need this and it looks like other people want this to

File: stable/redis-ha/values.yaml [hunk 1]
```diff
@@ -15,33 +15,14 @@ resources:
...
       memory: 200Mi
 
-## Node labels, tolerations, and affinities for pod assignment
+## Node labels and tolerations for pod assignment
 ## ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector
 ## ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#taints-and-tolerations-beta-feature
-## ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity
 nodeSelector: {}
 tolerations: []
-sentinelAffinity: |
```
Comment (ssalaues) [inline stable/redis-ha/values.yaml:24]: Curious, is there a specific reason you removed the explicit hard anti affinity?

Comment (RaymondArias) [reply stable/redis-ha/values.yaml:24]: I personally did not...I copied most of the pull request from @bzumhagan and made some fixes...so he likely removed it

**Comment by dung95tdn** [comment]
@RaymondArias Why did you delete the authentication in dockerfile ?

**Comment by RaymondArias** [comment]
@dung95 the contributor I took this over from deleted it

What I hoping will happen is that we can get the pr approved then we can add on the necessary feature we need.

Features that I know still need to be

- PVC for redis server
- Hard antiaffinity for sentinels
- Auth for redis pods

File: stable/redis-ha/templates/redis-server-statefulset.yaml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+          - name: REDIS_SENTINEL_SERVICE_HOST
+            value: "redis-sentinel"
+          - name: REDIS_CHART_PREFIX
+            value: {{ template "redis-ha.fullname" . }}-
+{{- if .Values.auth }}
+          - name: REDIS_PASS
+            valueFrom:
+              secretKeyRef:
+                name: {{ template "redis-ha.fullname" . }}
+                key: auth
```
Comment (RaymondArias) [inline stable/redis-ha/templates/redis-server-statefulset.yaml:62]: This looks correctly formatted to me

**Comment by RaymondArias** [comment]
@unguiculus and @mgoodness please review this

**Comment by RaymondArias** [comment]
Bug fixed

**Comment by RaymondArias** [comment]
@unguiculus and @mgoodness a review please?

**Comment by ssalaues** [comment]
@RaymondArias I was testing these changes in a kubespray style deployment to get a measure of the HA capabilities of the chart in general along with these changes. In testing I found this chart has lots of issues:

Race condition with masters: #5441
Announced service no longer working: #3197
PVCs possibilites: #3403
redis-ha doesn't failover properly: #2780

I encountered these other issues as well because in my testing failover will not usually happen if at all. This is especially true if a node with the master is down because of the use of labels and the state of pod labeled `master` becomes `Unknown` by Kubernetes design however retains the `master` label in turn causing all the other pods to believe that a master exists. Because of this and all the other issues I've also experienced, including the reasons why this PR was opened (Statefulsets), I decided to refactor this chart implementation of redis in PR #7323

TLDR; While having the statefulsets is a necessary change, there's still lots of fundamental issues with this chart and hopefully my PR addresses the needs of this one along with other issues that needed to be fixed. =/
